Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm

```diff
@@ -0,0 +1,305 @@
+
+    completionHandler(std::nullopt);
+}
+
+void WebExtensionContext::actionGetBadgeText(std::optional<WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebExtensionTabIdentifier> tabIdentifier, CompletionHandler<void(std::optional<String>, std::optional<String>)>&& completionHandler)
+{
+    static NSString * const apiName = @"action.getBadgeText()";
+
+    std::optional<String> error;
+    auto action = getActionWithIdentifiers(windowIdentifier, tabIdentifier, *this, apiName, error);
+    if (error) {
+        completionHandler(std::nullopt, error);
+        return;
+    }
+
+    ASSERT(action);
+
+    completionHandler(action->badgeText(), std::nullopt);
+}
+
+void WebExtensionContext::actionSetBadgeText(std::optional<WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebExtensionTabIdentifier> tabIdentifier, const String& text, CompletionHandler<void(std::optional<String>)>&& completionHandler)
+{
+    static NSString * const apiName = @"action.setBadgeText()";
+
+    std::optional<String> error;
+    auto action = getOrCreateActionWithIdentifiers(windowIdentifier, tabIdentifier, *this, apiName, error);
+    if (error) {
+        completionHandler(error);
+        return;
+    }
+
+    ASSERT(action);
+
+    action->setBadgeText(text);
+
+    completionHandler(std::nullopt);
+}
+
+void WebExtensionContext::actionGetEnabled(std::optional<WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebExtensionTabIdentifier> tabIdentifier, CompletionHandler<void(std::optional<bool>, std::optional<String>)>&& completionHandler)
+{
+    static NSString * const apiName = @"action.isEnabled()";
+
+    std::optional<String> error;
+    auto action = getActionWithIdentifiers(windowIdentifier, tabIdentifier, *this, apiName, error);
+    if (error) {
+        completionHandler(std::nullopt, error);
+        return;
+    }
+
+    ASSERT(action);
+
+    completionHandler(action->isEnabled(), std::nullopt);
+}
+
+void WebExtensionContext::actionSetEnabled(std::optional<WebExtensionTabIdentifier> tabIdentifier, bool enabled, CompletionHandler<void(std::optional<String>)>&& completionHandler)
+{
+    static NSString * const apiName = enabled ? @"action.enable()" : @"action.disable()";
+
+    std::optional<String> error;
+    auto action = getOrCreateActionWithIdentifiers(std::nullopt, tabIdentifier, *this, apiName, error);
+    if (error) {
+        completionHandler(error);
+        return;
+    }
+
+    ASSERT(action);
+
+    action->setEnabled(enabled);
+
+    completionHandler(std::nullopt);
+}
+
+void WebExtensionContext::fireActionClickedEventIfNeeded(WebExtensionTab* tab)
+{
+    constexpr auto type = WebExtensionEventListenerType::ActionOnClicked;
+    wakeUpBackgroundContentIfNecessaryToFireEvents({ type }, [&] {
+        sendToProcessesForEvent(type, Messages::WebExtensionContextProxy::DispatchActionClickedEvent(tab ? std::optional(tab->parameters()) : std::nullopt));
+    });
+}
```

Comment:
Is there a FIXME/bug about supporting badge text color?

Reply:
Yep you found it.